src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_GRPCSERVER_H_INCLUDED
+#define RIPPLE_CORE_GRPCSERVER_H_INCLUDED
+
+#include <ripple/rpc/Role.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/GRPCHandlers.h>
+#include <ripple/rpc/impl/RPCHelpers.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <ripple/rpc/impl/Tuning.h>
+#include <ripple/net/InfoSub.h>
+#include <ripple/protocol/ErrorCodes.h>
+#include <ripple/core/JobQueue.h>
+#include <ripple/resource/Charge.h>
+
+#include <grpcpp/grpcpp.h>
+#include "rpc/v1/xrp_ledger.grpc.pb.h"
+
+namespace ripple {
+
+//Interface that CallData implements
+class Processor
+{
+    public:
+
+    virtual ~Processor() {}
+
+    // process a request that has arrived. Can only be called once per instance
+    virtual void process() = 0;
+
+    //store an iterator to this object
+    //all Processor objects are stored in a std::list, and the iterator points
+    //to that objects position in the list. When object finishes processing a 
+    //request, the iterator is used to delete the object from the list
+    virtual void set_iter(
+            std::list<std::shared_ptr<Processor>>::iterator const& it) = 0;
+
+    //get iterator to this object. see above comment
+    virtual std::list<std::shared_ptr<Processor>>::iterator get_iter() = 0;
+
+    //abort processing this request. called when server shutsdown
+    virtual void abort() = 0;
+
+    //create a new instance of this CallData object, with the same type
+    //(same template parameters) as original. This is called when a CallData
+    //object starts processing a request. Creating a new instance allows the 
+    //server to handle additional requests while the first is being processed
+    virtual std::shared_ptr<Processor> clone() = 0;
+
+    //true if this object has finished processing the request. Object will be
+    //deleted once this function returns true
+    virtual bool isFinished() = 0;
+};
+
+namespace {
+
+//helper function. strips port from endpoint string
+std::string getEndpoint(std::string const& peer)
+{
+    std::size_t first = peer.find_first_of(":");
+    std::size_t last = peer.find_last_of(":");
+    std::string peer_clean(peer);
+    if(first != last)
+    {
+        peer_clean = peer.substr(first+1);
+    }
+    return peer_clean;
+}
+
+} //namespace
+
+class GRPCServerImpl final {
+
+    private:
+
+    //list of current RPC requests being processed or listened for
+    std::list<std::shared_ptr<Processor>> requests_;
```

Comment:
Can we remove this member function and move the collection to `GRPCServerImpl::handleRpcs()
`?

I'd also make this collection a `std::vector<std::unique_ptr<Processor>>`  (assuming this controls the lifetime of the Processors). I'm already suggesting that we don't store the iterator in the processor object, so there are no iterator invalidation issues when changing to a vector.

Reply:
Why you would use a vector when objects could be deleted from any position in the collection? Deleting from a vector is an `O(n)` operation (especially when deleting from the front or middle), whereas deleting from a linked list is `O(1)` (if you know the position). Most of the time, the objects to be deleted would be towards the front of the vector, as requests will finish somewhat close to the order in which they arrived. So, using a vector would have asymptotically worse performance than using a linked list.

We can move the collection to `handleRpcs()` though; I think that makes more sense than storing the collection as a member.

Reply:
@cjcobb23 If we don't care about the order of the vector (and if we don't store the iterator, I don't think we care - please correct me if I'm wrong) you can delete in O[1] by swapping the element you want to delete with the last element in the vector, and then shrink the vector by one. There's no need to move all the elements, just swap two (which should be very fast).

Reply:
Ah that's a good idea with the swapping. I'll do that.

About removing the iterator, I would like to remove it as well, and just search the vector to find the element. Just an FYI, since `cq_.Next()` sets the `tag` variable to a raw pointer, and the vector will contain `shared_ptr`, to find the element to delete, I'm going to have to either:
 1. call `get()` on each `shared_ptr` as I traverse the vector, and compare to the raw pointer.
 2. use `shared_from_this()` to get a `shared_ptr` that can be passed to `std::find`.

I prefer 1, unless you can think of a better way.

Reply:
@cjcobb23 I think (1) is fine.

Reply:
@seelabs About your original request of moving the `requests_` collection to `handleRpcs()`: we need access to that collection in `setupListeners()`, where we create a `CallData` object for each RPC, and add it to the collection.

Related, that setup logic needs to execute on the main application thread (`handleRpcs()` is running in its own dedicated thread), else the unit tests can fail because the gRPC server is not setup at the time that the unit tests try to execute the gRPC requests. Maybe there is a better way to write the unit tests such that they wait for the gRPC server to setup, but as the code is now, `setupListeners()` needs to happen on the main thread. Take a look at `GRPCServer::run()` to see what I'm talking about.

Reply:
@cjcobb Here's a gist of what I had in mind for `handleRpcs: https://gist.github.com/seelabs/87ff42e3a7e7023d7278a1b46c24cb1f

So `request` is local to `handleRpcs` and we set it up the collection prior to the loop in that function. That doesn't work? The unit tests seem to run OK here when I do that. It's very possible that I'm confused about something and this won't work (you're much more familiar with gRPC than I am).

Reply:
@seelabs Actually I was mistaken. The server itself must be started in the main thread (everything that happens in `GRPCServerImpl::start()` besides `setupListeners()`), but the `CallData` objects can be constructed in the dedicated thread. So, I will make `requests` local to `handleRpcs()`

Reply:
@seelabs I think your gist looks fine. I'm ok with removing the `makeAndPush()` helper function and just using the lambda instead. I would like to move the instantiation of the `CallData` objects to `setupListeners()` though, like here: [https://gist.github.com/cjcobb23/d61d4f115c267df5faca174a17b3fc3e](https://gist.github.com/cjcobb23/d61d4f115c267df5faca174a17b3fc3e) . There are only 4 `CallData` objects to instantiate now, but as we add more RPCs, there could be a lot more.

Reply:
@cjcobb23 I'm fine with moving the setup of the collection to another function, tho I'd just have that function return the collection as a result, no need to pass it as a parameter.

Reply:
changed this. `requests` is returned by `setupListeners()` as opposed to being a member variable.